File: codes/scgpap/haversines/haversine.py
from math import sin,cos,radians,asin,sqrt
from time import time
from random import random

earth_radius_km = 6371.0
N = 10000000

rnd = lambda max: random() * max
args = []
for i in range(N):
	args.append(rnd(360.0)-180.0)
	args.append(rnd(180.0)-90.0)
	args.append(rnd(360.0)-180.0)
	args.append(rnd(180.0)-90.0)

def haversine_of_degrees(x0,y0,x1,y1,r):
	dy = radians(y1-y0)
	dx = radians(x1-x0)
	y0r = radians(y0)
	y1r = radians(y1)
	syroot = sin(dy * 0.5)
	sxroot = sin(dx * 0.5)
	root_term = syroot*syroot + cos(y0r)*cos(y1r)*sxroot*sxroot
	return 2.0 * r * asin(sqrt(root_term))


N_TRIALS=10
best_rate = 0
for i0 in range(N_TRIALS):
	sum = 0.0
	offset = 0
	t0 = time()
	for i in range(N):
		sum += haversine_of_degrees(args[offset+0], args[offset+1], args[offset+2], args[offset+3], earth_radius_km)
		offset += 4
	dt = time() - t0
	rate = float(N)/dt
	if rate > best_rate: best_rate = rate
	print("avg is %.2f; %.0f haversines/s" % (sum/float(N), rate))
print("best: %.0f haversines/s" % best_rate)

File: codes/scgpap/haversines/haversine.js
const earth_radius_km = 6371;
const N = 10000000;

const rnd = (max) => Math.random() * max;
/*
let args = [];
for (let i = 0; i < N; i++) {
	args.push(rnd(360)-180);
	args.push(rnd(180)-90);
	args.push(rnd(360)-180);
	args.push(rnd(180)-90);
}
*/
let args = new Float64Array(N*4);
for (let offset = 0; offset < N*4; offset+=4) {
	args[offset+0] = rnd(360)-180;
	args[offset+1] = rnd(180)-90;
	args[offset+2] = rnd(360)-180;
	args[offset+3] = rnd(180)-90;
}

const radians = (degrees) => 0.017453292519943295 * degrees;

const haversine_of_degrees = (x0,y0,x1,y1,r) => {
	const dy = radians(y1-y0);
	const dx = radians(x1-x0);
	const y0r = radians(y0);
	const y1r = radians(y1);
	const syroot = Math.sin(dy * 0.5);
	const sxroot = Math.sin(dx * 0.5);
	const root_term = syroot*syroot + Math.cos(y0r)*Math.cos(y1r)*sxroot*sxroot;
	return 2.0 * r * Math.asin(Math.sqrt(root_term));
};

const N_TRIALS = 10;
let best_rate = 0;
for (let i0 = 0; i0 < N_TRIALS; i0++) {
	let sum = 0.0;
	let offset = 0;
	let t0 = Date.now();
	for (let i = 0; i < N; i++) {
		sum += haversine_of_degrees(args[offset+0], args[offset+1], args[offset+2], args[offset+3], earth_radius_km);
		offset += 4;
	}
	const dt = Date.now() - t0;
	const rate = (N/(dt/1000.0));
	if (rate > best_rate) best_rate = rate;
	console.log("avg is " + (sum/N).toFixed(0) + ". " + (N/(dt/1000.0)).toFixed(2) + " haversines/s");
}
console.log("best: " + best_rate.toFixed(0) + " haversines/s");
